server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -260,7 +243,8 @@ public void recoverToTarget(ActionListener<RecoveryResponse> listener) {
                     sendSnapshotResult.totalOperations, sendSnapshotResult.tookTime.millis())
             );
         } catch (Exception e) {
-            onFailure.accept(e);
+            IOUtils.closeWhileHandlingException(resources);
```

Comment:
I'd feel better if  we  would do it this way ` IOUtils.closeWhileHandlingException(resources, () -> wrappedListener.onFailure(e));`

Reply:
+1. I pushed https://github.com/elastic/elasticsearch/pull/37291/commits/ce141399cbd08578e018a33bbdd77ed9b8556aac.